rust/src/mysql/parser.rs

```diff
@@ -0,0 +1,3350 @@
+                Ok((
+                    i,
+                    MysqlResponse {
+                        item: MysqlResponsePacket::Unknown,
+                    },
+                ))
+            }
+        }
+    })(i)?;
+    Ok((
+        i,
+        MysqlResponse {
+            item: MysqlResponsePacket::ResultSet {
+                n_cols,
+                columns,
+                eof,
+                rows,
+            },
+        },
+    ))
+}
+
+fn parse_response_binary_resultset(i: &[u8], n_cols: u64) -> IResult<&[u8], MysqlResponse> {
+    let (i, columns) = many_m_n(n_cols as usize, n_cols as usize, parse_column_definition)(i)?;
+    let (i, eof) = parse_eof_packet(i)?;
+    let (i, (rows, _)) = many_till(parse_binary_resultset_row(columns), |i| {
+        // eof
+        let (i, header) = parse_packet_header(i)?;
+        let (i, response_code) = verify(be_u8, |&x| x == 0xFE || x == 0xFF)(i)?;
+        match response_code {
+            // EOF
+            0xFE => {
+                let (i, _) = cond(header.pkt_len - 1 > 0, take(header.pkt_len - 1))(i)?;
+                Ok((
+                    i,
+                    MysqlResponse {
+                        item: MysqlResponsePacket::EOF,
+                    },
+                ))
+            }
+            // ERR
+            0xFF => parse_response_err(i, header.pkt_len - 1),
+            _ => {
+                let (i, _) = cond(header.pkt_len - 1 > 0, take(header.pkt_len - 1))(i)?;
+                Ok((
+                    i,
+                    MysqlResponse {
+                        item: MysqlResponsePacket::Unknown,
+                    },
+                ))
+            }
+        }
+    })(i)?;
+    Ok((
+        i,
+        MysqlResponse {
+            item: MysqlResponsePacket::BinaryResultSet { n_cols, eof, rows },
+        },
+    ))
+}
+
+fn parse_response_ok(i: &[u8], length: u32) -> IResult<&[u8], MysqlResponse> {
+    let old = i;
+    let (i, rows) = parse_varint(i)?;
+    let (i, _last_insert_id) = parse_varint(i)?;
+    let (i, flags) = le_u16(i)?;
+    let (i, warnings) = le_u16(i)?;
+    let consumed = (old.len() - i.len()) as u32;
+    // Should never happen
+    if consumed > length {
+        return Ok((
+            &[],
+            MysqlResponse {
+                item: MysqlResponsePacket::Ok {
+                    rows,
+                    flags,
+                    warnings,
+                },
+            },
+        ));
+    }
+    let (i, _) = take(length - consumed)(i)?;
+
+    Ok((
+        i,
+        MysqlResponse {
+            item: MysqlResponsePacket::Ok {
+                rows,
+                flags,
+                warnings,
+            },
+        },
+    ))
+}
+
+fn parse_response_err(i: &[u8], length: u32) -> IResult<&[u8], MysqlResponse> {
+    let (i, error_code) = le_u16(i)?;
+    let (i, _) = take(6_u32)(i)?;
+    // sql state maker & sql state
+    let (i, _) = take(6_u32)(i)?;
```

Comment:
sql state marker and state are available only if `capabilities & CLIENT_PROTOCOL_41` is true.

Reply:
When i parse handshake response, i have checked this.